W 12305455 flex liveness probe gr (#261)
mulesoft/docs-gateway: gateway/1.4/modules/ROOT/pages/_partials/task-config-liveness-probe.adoc

//tag::intro1[]

Flex Gateway includes the following CLI command to test gateway health. You can run this command manually on any Flex Gateway:

[source,ssh]
----
flexctl probe --check=liveness
----

This command returns either an exit code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.

//end::intro1[]

//tag::intro2[]
You can also configure this command to run automatically. By default, Flex Gateways deployed on Kubernetes have a liveness probe configured. The liveness probe periodically runs the liveness check command and automatically restarts the Flex Gateway pod after a specified number of failed liveness checks. 

> Flex Gateways deployed on Docker or Linux, Flex Gateway does not offer a default liveness probe. However, you can still use the the liveness check command. For both deployments, you can use liveness check in scripts that check liveness and restart containers. 

Reviewer: IMHO:
- I would avoid saying that "Flex Gateway does not offer a default liveness probe." because I'm afraid that users might think that they don't have a way to configure a liveness probe. Maybe something like "Flex Gateway does not configure a default liveness probe."
- I would avoid using the word "script" here, since liveness probes can be configured through other ways too

```suggestion
Flex Gateways deployed on Docker or Linux do not configure a default liveness probe. However, you can still use the the liveness check command. For both deployments, you can use liveness check to check liveness and restart Flex Gateway. 
```

//tag::intro1[]

> Flex Gateway includes the following CLI command to test gateway health. You can run this command manually on any Flex Gateway:

Reviewer 2: ```suggestion
Flex Gateway includes the following CLI command to test gateway health:
```
Not mentioning that they can also run the command automatically implies it's not possible to do so, at least at first glance. I think it's better to omit mention to manually running, and mention both together below (see another suggestion.)

[source,ssh]
----
flexctl probe --check=liveness
----

> This command returns either an exit code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.

Author: ```suggestion
If the Flex Gateway instance is healthy, the command returns an exit code of `0`. Otherwise, it returns a `1`.
```

> This command returns either an exit code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.

Reviewer 2: ```suggestion
If the Flex Gateway instance is healthy, the command returns an exit code of `0`. Otherwise it returns a `1`.
```
Shortening sentences.

//end::intro1[]

//tag::intro2[]

> You can also configure this command to run automatically. By default, Flex Gateways deployed on Kubernetes have a liveness probe configured. The liveness probe periodically runs the liveness check command and automatically restarts the Flex Gateway pod after a specified number of failed liveness checks. 

Reviewer 2: ```suggestion
You can either run this command manually or configure the command to run automatically. By default, Flex Gateways deployed on Kubernetes have an automatic liveness probe configured. The liveness probe periodically runs the liveness check command and automatically restarts the Flex Gateway pod after a specified number of failed liveness checks. 
```

> For Docker or Linux deployments, there is not a default liveness probe configured to check health or restart unhealthy Flex Gateways. However, you can still use the the liveness check command manually or configure it for automated use with third-party services. Besides the Kubernetes liveness probe, Mulesoft does not provide support for third-party liveness probes or health checks.

Reviewer 2: ```suggestion
For Docker or Linux deployments, there is no default liveness probe configured to check health or restart unhealthy Flex Gateways. However, you can still run the the liveness check command manually or configure it for automated use with third-party services. Besides the Kubernetes liveness probe, MuleSoft does not provide support for third-party liveness probes or health checks.
```

One method of running the liveness check command with Docker is to configure `HEALTHCHECK` in your `docker run` command. For more information, see https://docs.docker.com/engine/reference/run/#healthcheck["Docker run HEALTHCHECK"].

> //tag::intro2[]

Reviewer 3: ```suggestion
//end::intro2[]
```

> An example of how you can use the liveness check command with Docker is by configuring `HEALTHCHECK` in your `docker run` command. For more information about `HEALTHCHECK`, see https://docs.docker.com/engine/reference/run/#healthcheck[Docker run HEALTHCHECK].

Reviewer 2: ```suggestion
One method of running the liveness check command with Docker is to configure `HEALTHCHECK` in your `docker run` command. For more information, see https://docs.docker.com/engine/reference/run/#healthcheck["Docker run HEALTHCHECK"].
```

//tag::k8s-liveness-probe[]
== Before You Begin

> Register and deploy Flex Gateway to a Kubernetes cluster. For guidance, see xref:flex-gateway-set-up.adoc[] or xref:flex-gateway-k8-getting-started.adoc[].

Reviewer 2: ```suggestion
Register a Flex Gateway instance and deploy this instance to a Kubernetes cluster. For guidance, see xref:flex-gateway-set-up.adoc[] or xref:flex-gateway-k8-getting-started.adoc[].
```
Rewording for clarity (it seems like you're saying they should register to the cluster, at least this was my initial read of it.)

== Configure a Kubernetes Liveness Probe

> By default, Flex Gateways running on Kubernetes have a liveness probe configured in the Helm chart. The liveness probe automatically runs the liveness check command at a specified interval and restarts unhealthy pods after a specified number of unhealthy tests. 

Reviewer 2: ```suggestion
By default, Flex Gateways running on Kubernetes include a liveness probe configured in the Helm chart. The liveness probe automatically runs `flexctl probe --check=liveness` at a specified interval and restarts unhealthy pods after a specified number of failed tests. 
```
Probably good to specify the exact command so that they don't have to scroll back up the page to see it.

The liveness probe is configured by default with the following values:

[source,helm]
----
livenessProbe:
 exec:
   command:
   - flexctl
   - probe
   - --check=liveness
 initialDelaySeconds: 10
 periodSeconds: 10
 failureThreshold: 5
 timeoutSeconds: 1
----

[cols="1,2"]
|===
| Parameter | Description

| `exec.command` | The liveness check command
| `initialDelaySeconds` | The time in seconds to wait after startup before executing the first liveness test
| `periodSeconds` | The period in seconds between each liveness check
| `failureThreshold` | The number of failed liveness tests before the Kubernetes pod is restarted
| `timeoutSeconds` | The time in seconds before the test times out
|===

> To modify the default liveness probe parameters, you must update your Helm chart. You can modify these parameters at the initial installation of the Helm chart or after. For more information about how to modify a Helm chart, see xref:flex-gateway-k8-change-helm-settings.adoc[]

Reviewer 3: ```suggestion
To modify the default liveness probe parameters, you must update your Helm chart. You can modify these parameters before or after the initial installation of the Helm chart. For more information about how to modify a Helm chart, see xref:flex-gateway-k8-change-helm-settings.adoc[]
```

//tag::intro1[]

> Flex Gateway offers the following CLI command to test the health of a Flex Gateway:

Reviewer 2: ```suggestion
Flex Gateway includes the following CLI command to test gateway health:
```
Trimming for wordiness.

[source,ssh]
----
flexctl probe --check=liveness
----

> This command returns either a status code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.

Reviewer: ```suggestion
This command returns either an exit code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.
```

> This command returns either a status code of `0` for a healthy Flex Gateway or `1` for an unhealthy Flex Gateway.

Reviewer 2: ```suggestion
If the Flex Gateway instance is healthy, the command returns a status code of `0`. Otherwise it returns a `1`.
```
Shortening sentences.

//end::intro1[]

//tag::intro2[]

> By using the liveness check command, you can either configure a Kubernetes liveness probe or a Docker health check command to restart a failed Flex Gateway.

Reviewer: A couple of things to here:
- From 1.4.4 we will configure a default liveness probe in kubernetes that can be customized by the users
- In docker, the health check only marks the container as healthy or unhealthy, but it doesn't restart it

```suggestion
By using the liveness check command, you can either customize the Kubernetes liveness probe provided by default to restart failed Flex Gateway pods or configure a Docker health check command.
```

> By using the liveness check command, you can either configure a Kubernetes liveness probe or a Docker health check command to restart a failed Flex Gateway.

Reviewer 2: ```suggestion
The liveness check command enables you to configure a Kubernetes liveness probe or a Docker health check. Either can be configured to restart a failed Flex Gateway.
```
Shortening sentences.

> For Flex Gateways running on Linux, you can still run the liveness check command. However, there is not an easy to automate the Linux machine to restart when the command returns an unhealthy status.

Reviewer: IMHO, I would try to avoid saying that theres not an easy way to automate the restart of the process. There might be expert Linux users reading this that know how to do it. Also, I suggest replacing "the command returns an unhealthy status" with "when the check/probe fails"

```suggestion
For Flex Gateways running on Linux, you can still run the liveness check command manually. However, we don't provide an automated way to restart the process when the check fails.
```

> For Flex Gateways running on Linux, you can still run the liveness check command. However, there is not an easy to automate the Linux machine to restart when the command returns an unhealthy status.

Reviewer 2: Should we even mention this if we're not going to describe how to configure a check for Linux? Is Linux officially not supported? It seems strange to only say there's no easy way to do it. Customers will undoubtedly wonder about the "hard way", and why it's not mentioned.

//tag::intro2[]

//tag::k8s-liveness-probe[]
== Configure a Kubernetes Liveness Probe

> For Flex Gateways running on Kubernetes, you can configure a Kubernetes liveness probe to monitor and restart a Flex Gateway pod when unhealthy. To learn more about Kubernetes probes, see https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/[Configure Liveness, Readiness and Startup Probes^].

Reviewer 2: ```suggestion
For Flex Gateways running on Kubernetes, you can configure a liveness probe to monitor and restart an unhealthy pod. To learn more about Kubernetes probes, see https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/[Configure Liveness, Readiness and Startup Probes^].
```
Removing some redundancy.

> To enable a liveness probe, update you Helm chart with the following configuration:

Reviewer: Here we need to mention that users can customize the probe, but that one is provided by default

> To enable a liveness probe, update you Helm chart with the following configuration:

Reviewer 2: ```suggestion
To enable a liveness probe, update your Helm chart with the following configuration:
```

[source,helm]
----
livenessProbe:
 exec:
   command:
   - flexctl
   - probe
   - --check=liveness
 initialDelaySeconds: 10
 periodSeconds: 10
 failureThreshold: 5
 timeoutSeconds: 1
----

[cols="1,2"]
|===
| Parameter | Description

> | `exec.command` | The command to execute to test liveness

Reviewer 2: This is a little confusing, maybe because there are typos? Is it the command that tests liveliness?

| `initialDelaySeconds` | The time in seconds to wait after startup before executing the first liveness test
| `periodSeconds` | The period in seconds between each liveness check
| `failureThreshold` | The number of failed liveness tests before the Kubernetes pod is restarted
| `timeoutSeconds` | The time in seconds before the test times out
|===

> To learn more about updating a Helm chart, see xref:gateway::flex-gateway-k8-change-helm-settings.adoc[Update a Helm Chart].

Reviewer 2: ```suggestion
To learn more about updating a Helm chart, see xref:flex-gateway-k8-change-helm-settings.adoc[Update a Helm Chart].
```
Use relative paths in xrefs.

//end::k8s-liveness-probe[]

//tag::docker-liveness-probe[]
== Docker Liveness Probe

For Flex Gateways running on Docker, you can configure a health check using flags in the `docker run` command.

> After adding the relative Flex Gateway registration information from xref:gateway::flex-{page-mode}-reg-run.adoc[], execute the `docker run` command with the following flags configured:

Reviewer 2: ```suggestion
After adding the relative Flex Gateway registration information from xref:flex-{page-mode}-reg-run.adoc[], execute the `docker run` command with the following flags configured:
```
Use relative paths in xrefs.

Also, what is "relative Flex Gateway registration information"? Is the user supposed to copy the docker registration command and use it below?

[source,ssh]
----
docker run --health-retries=5 \
--health-interval=10s \
--health-start-period=10s \
--health-timeout=1s \
--health-cmd="flexctl probe --check=liveness" \
--entrypoint flexctl \
-v "$(pwd)":/registration -u $UID mulesoft/flex-gateway \
register \

> #REPLACE WITH RELATIVE FLEX GATEWAY REGISTRATION

Reviewer 2: I think this is a little confusing. Is "relative" the correct word? Do you mean "related"? Are they copying a registration command from elsewhere?

> my-gateway

Reviewer: Actually, the command that we should use here is the one to start flex, not to register https://docs.mulesoft.com/gateway/latest/flex-local-reg-run-up#start-command

----

[cols="1,2"]
|===
| Flag | Description

| `--health-cmd` | The command to execute to test liveness
| `--health-start-period` | The time in seconds to wait after startup before executing the first liveness test
| `--health-interval` | The period in seconds between each liveness check

> | `--health-retries` | The number of failed liveness tests before the Kubernetes pod is restarted

Reviewer 2: Is this the Docker section? Should it refer to Kubernetes pods here?

| `--health-timeout` | The time in seconds before the test times out
|===

//end::docker-liveness-probe[]

//tag::see-also[]
== See Also

> * xref:gateway::policies-included-health-check.adoc[]

Reviewer 2: ```suggestion
* xref:policies-included-health-check.adoc[]
```
Use relative paths in xrefs.
